flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogBase.java

```diff
@@ -407,4 +443,201 @@ private Table getHiveTable(ObjectPath tablePath) throws TableNotExistException {
 				String.format("Failed to get table %s from Hive metastore", tablePath.getFullName()), e);
 		}
 	}
+
+	// ------ partitions ------
+
+	@Override
+	public boolean partitionExists(ObjectPath tablePath, CatalogPartitionSpec partitionSpec)
+		throws CatalogException {
+		try {
+			Table hiveTable = getHiveTable(tablePath);
+			return client.getPartition(tablePath.getDatabaseName(), tablePath.getObjectName(),
+				getFullPartitionValues(tablePath, partitionSpec, getFieldNames(hiveTable.getPartitionKeys()))) != null;
+		} catch (NoSuchObjectException | TableNotExistException | PartitionSpecInvalidException e) {
+			return false;
+		} catch (TException e) {
+			throw new CatalogException(
+				String.format("Failed to get partition %s of table %s", partitionSpec, tablePath), e);
+		}
+	}
+
+	@Override
+	public void createPartition(ObjectPath tablePath, CatalogPartitionSpec partitionSpec, CatalogPartition partition, boolean ignoreIfExists)
+		throws TableNotExistException, TableNotPartitionedException, PartitionSpecInvalidException, PartitionAlreadyExistsException, CatalogException {
+
+		validateCatalogPartition(partition);
+
+		Table hiveTable = getHiveTable(tablePath);
+
+		ensurePartitionedTable(tablePath, hiveTable);
+
+		try {
+			client.add_partition(createHivePartition(hiveTable, partitionSpec, partition));
+		} catch (AlreadyExistsException e) {
+			if (!ignoreIfExists) {
+				throw new PartitionAlreadyExistsException(catalogName, tablePath, partitionSpec);
+			}
+		} catch (TException e) {
+			throw new CatalogException(
+				String.format("Failed to create partition %s of table %s", partitionSpec, tablePath));
+		}
+	}
+
+	@Override
+	public void dropPartition(ObjectPath tablePath, CatalogPartitionSpec partitionSpec, boolean ignoreIfNotExists)
+		throws PartitionNotExistException, CatalogException {
+		try {
+			Table hiveTable = getHiveTable(tablePath);
+			ensurePartitionedTable(tablePath, hiveTable);
+			client.dropPartition(tablePath.getDatabaseName(), tablePath.getObjectName(),
+				getFullPartitionValues(tablePath, partitionSpec, getFieldNames(hiveTable.getPartitionKeys())), true);
+		} catch (NoSuchObjectException e) {
+			if (!ignoreIfNotExists) {
+				throw new PartitionNotExistException(catalogName, tablePath, partitionSpec);
+			}
+		} catch (TException e) {
+			throw new CatalogException(
+				String.format("Failed to drop partition %s of table %s", partitionSpec, tablePath));
+		} catch (TableNotPartitionedException | TableNotExistException | PartitionSpecInvalidException e) {
+			throw new PartitionNotExistException(catalogName, tablePath, partitionSpec, e);
+		}
+	}
+
+	@Override
+	public List<CatalogPartitionSpec> listPartitions(ObjectPath tablePath)
+		throws TableNotExistException, TableNotPartitionedException, CatalogException {
+		Table hiveTable = getHiveTable(tablePath);
+
+		ensurePartitionedTable(tablePath, hiveTable);
```

Comment:
Similar question as above.

Reply:
Same as `createPartition`, this API requires more specific exception. Better not to rely on HMS client for it.